File: siden-core/src/main/java/ninja/siden/jmx/WebSocketTracker.java
/*
 * Copyright 2015 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.jmx;

import java.nio.ByteBuffer;

import ninja.siden.Connection;
import ninja.siden.WebSocket;
import ninja.siden.WebSocketFactory;

/**
 * @author taichi
 */
public class WebSocketTracker implements WebSocketFactory, WebSocketMXBean {

	final WebSocketFactory original;

	RequestMeter onConnect = new RequestMeter();
	RequestMeter onText = new RequestMeter();
	RequestMeter onBinary = new RequestMeter();
	RequestMeter onPong = new RequestMeter();
	RequestMeter onPing = new RequestMeter();
	RequestMeter onClose = new RequestMeter();

	public WebSocketTracker(WebSocketFactory original) {
		this.original = original;
	}

	@Override
	public WebSocket create(Connection connection) {
		return new WsWrapper(this.original.create(connection));
	}

	class WsWrapper implements WebSocket {
		final WebSocket original;

		public WsWrapper(WebSocket original) {
			this.original = original;
		}

		@Override
		public void onConnect(Connection connection) throws Exception {
			onConnect.accept(m -> original.onConnect(connection));
		}

		@Override
		public void onText(String payload) throws Exception {
			onText.accept(m -> original.onText(payload));
		}

		@Override
		public void onBinary(ByteBuffer[] payload) throws Exception {
			onBinary.accept(m -> original.onBinary(payload));
		}

		@Override
		public void onPong(ByteBuffer[] payload) throws Exception {
			onPong.accept(m -> original.onPong(payload));
		}

		@Override
		public void onPing(ByteBuffer[] payload) throws Exception {
			onPing.accept(m -> original.onPing(payload));
		}

		@Override
		public void onClose(ByteBuffer[] payload) throws Exception {
			onClose.accept(m -> original.onClose(payload));
		}
	}

	@Override
	public void reset() {
		this.onConnect.reset();
		this.onText.reset();
		this.onBinary.reset();
		this.onPong.reset();
		this.onPing.reset();
		this.onClose.reset();
	}

	@Override
	public RequestMetrics getOnConnect() {
		return onConnect.toMetrics();
	}

	@Override
	public RequestMetrics getOnText() {
		return onText.toMetrics();
	}

	@Override
	public RequestMetrics getOnBinary() {
		return onBinary.toMetrics();
	}

	@Override
	public RequestMetrics getOnPong() {
		return onPong.toMetrics();
	}

	@Override
	public RequestMetrics getOnPing() {
		return onPing.toMetrics();
	}

	@Override
	public RequestMetrics getOnClose() {
		return onClose.toMetrics();
	}
}

File: siden-core/src/main/kotlin/ninja/siden/jmx/WebSocketTracker.kt
/*
 * Copyright 2015 SATO taichi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ninja.siden.jmx

import ninja.siden.Connection
import ninja.siden.WebSocket
import ninja.siden.WebSocketFactory

import java.nio.ByteBuffer

/**
 * @author taichi
 */
class WebSocketTracker(internal val original: WebSocketFactory) : WebSocketFactory, WebSocketMXBean {

    internal var _onConnect = RequestMeter()
    internal var _onText = RequestMeter()
    internal var _onBinary = RequestMeter()
    internal var _onPong = RequestMeter()
    internal var _onPing = RequestMeter()
    internal var _onClose = RequestMeter()

    override fun create(connection: Connection): WebSocket {
        return WsWrapper(this.original.create(connection))
    }

    internal inner class WsWrapper(val original: WebSocket) : WebSocket {
        override fun onConnect(connection: Connection) = _onConnect.record { original.onConnect(connection) }
        override fun onText(payload: String) = _onText.record { original.onText(payload) }
        override fun onBinary(payload: Array<ByteBuffer>) = _onBinary.record { original.onBinary(payload) }
        override fun onPong(payload: Array<ByteBuffer>) = _onPong.record { original.onPong(payload) }
        override fun onPing(payload: Array<ByteBuffer>) = _onPing.record { original.onPing(payload) }
        override fun onClose(payload: Array<ByteBuffer>) = _onClose.record { original.onClose(payload) }
    }

    override fun reset() {
        this._onConnect.reset()
        this._onText.reset()
        this._onBinary.reset()
        this._onPong.reset()
        this._onPing.reset()
        this._onClose.reset()
    }

    override val onConnect: RequestMetrics
        get() = _onConnect.toMetrics()
    override val onText: RequestMetrics
        get() = _onText.toMetrics()
    override val onBinary: RequestMetrics
        get() = _onBinary.toMetrics()
    override val onPong: RequestMetrics
        get() = _onPong.toMetrics()
    override val onPing: RequestMetrics
        get() = _onPing.toMetrics()
    override val onClose: RequestMetrics
        get() = _onClose.toMetrics()
}
